battleground_viewer/src/construct_render/render.rs

```diff
@@ -0,0 +1,408 @@
+        };
+        material_new(
+            context,
+            &CpuMaterial {
+                albedo: Color {
+                    r: 255,
+                    g: 255,
+                    b: 255,
+                    a: 255,
+                },
+                ..Default::default()
+            },
+        )
+    }
+}
+
+impl BatchMaterial for ColorMaterial {
+    fn new_for_batch(context: &Context, batch_properties: BatchProperties) -> ColorMaterial {
+        let material_new = match batch_properties {
+            BatchProperties::None => ColorMaterial::new_opaque,
+            BatchProperties::Basic {
+                is_transparent: false,
+            } => ColorMaterial::new_opaque,
+            BatchProperties::Basic {
+                is_transparent: true,
+            } => ColorMaterial::new_transparent,
+        };
+        material_new(
+            context,
+            &CpuMaterial {
+                albedo: Color {
+                    r: 255,
+                    g: 255,
+                    b: 255,
+                    a: 255,
+                },
+                ..Default::default()
+            },
+        )
+    }
+}
+
+pub struct MeshGeometry<M: Material + BatchMaterial> {
+    participates_in_pass: fn(RenderPass) -> bool,
+    buffer: Vec<(CpuMesh, Mat4, Color)>,
+    meshes: Vec<Gm<InstancedMesh, M>>,
+}
+
+impl<M: Material + BatchMaterial> MeshGeometry<M> {
+    pub fn new(participates_in_pass: fn(RenderPass) -> bool) -> Self {
+        Self {
+            participates_in_pass,
+            buffer: Default::default(),
+            meshes: Default::default(),
+        }
+    }
+
+    fn meshes_for_pass(
+        &self,
+        pass: RenderPass,
+    ) -> Option<impl Iterator<Item = &Gm<InstancedMesh, M>>> {
+        if (self.participates_in_pass)(pass) {
+            Some(self.meshes.iter())
+        } else {
+            None
+        }
+    }
+
+    pub fn add_mesh(&mut self, mesh: &CpuMesh, transform: Mat4, color: Color) {
+        self.buffer.push((mesh.clone(), transform, color));
+    }
+}
+
+impl<M: Material + BatchMaterial> RenderableGeometry for MeshGeometry<M> {
+    fn objects(&self, pass: RenderPass) -> Option<Vec<&dyn Object>> {
+        self.meshes_for_pass(pass)
+            .map(|xs| xs.map(|x| x as &dyn Object).collect::<_>())
+    }
+
+    fn geometries(&self, pass: RenderPass) -> Option<Vec<&InstancedMesh>> {
+        self.meshes_for_pass(pass)
+            .map(|xs| xs.map(|x| &x.geometry).collect::<_>())
+    }
+
+    fn prepare_frame(&mut self) {
+        self.buffer.clear();
+        self.meshes.clear();
+    }
+
+    fn finish_frame(&mut self, context: &Context) {
+        for (mesh, transform, color) in &self.buffer {
+            let instanced = Gm::new(
+                InstancedMesh::new(
+                    context,
+                    &Instances {
+                        transformations: vec![*transform],
+                        colors: Some(vec![*color]),
```

Comment:
I'm confused, what is the benefit of `InstancedMeshes` if we are only ever adding one instance? 

You might as well put them all into [`Gm`](https://docs.rs/three-d/latest/three_d/renderer/object/struct.Gm.html) holding your material and a [`Mesh`](https://docs.rs/three-d/latest/three_d/renderer/geometry/struct.Mesh.html) at this point? Or is the only benefit that we don't have to transform the actual mesh and can position it with the transformation matrix?

I presume this relates to the
>  This also means that it is not currently possible to "N of the same arbitrary mesh" to the world without peformance issues.

Statement from the description? From the description it isn't actually clear what the actual change is that requires this and why it is necessary or beneficial? Is it to avoid keying on the primitives? Because we still have the hash calculation for that...

I'm probably missing something, it's pretty late, I'll take a fresh look in the morning. Edit: Oh, `PrimitiveGeometry` still has uses the instancing.. What's `MeshGeometry` used for?

Reply:
A documentation line for `PrimitiveGeometry` and `MeshGeometry` that describes when each is used perhaps also useful?

Reply:
Ah, I see, well lets leave as is :+1: